Python/ceval_gil.c

```diff
@@ -966,6 +966,51 @@ _PyEval_InitState(PyInterpreterState *interp)
     _gil_initialize(&interp->_gil);
 }
 
+#ifdef Py_GIL_DISABLED
+int
+_PyEval_EnableGIL(PyThreadState *tstate)
+{
+    struct _gil_runtime_state *gil = tstate->interp->ceval.gil;
+
+    // gil->enabled only transitions from 0 to 1, and only while the world is
+    // stopped, so this can be read without any synchronization.
+    if (gil->enabled) {
+        return 0;
+    }
+
+    // Enabling the GIL changes what it means to be an "attached" thread. To
+    // safely make this transition, we:
+    // 1. Detach the current thread.
+    // 2. Stop the world to detach (and suspend) all other threads.
+    // 3. Enable the GIL, if nobody else did between our check above and when
+    //    our stop-the-world begins.
+    // 4. Start the world.
+    // 5. Attach the current thread. Other threads may attach and hold the GIL
+    //    before this thread, which is harmless.
+    _PyThreadState_Detach(tstate);
```

Comment:
As I was working out how to safely *disable* the GIL, I realized that this sequence to enable the GIL isn't safe:
1. Thread A is detached and calls `_PyThreadState_Attach()`.
2. [This call to `_PyEval_AcquireLock()`](https://github.com/python/cpython/blob/42205143f8b3211d1392f1d9f2cf6717bdaa5b47/Python/pystate.c#L1967) is a no-op in thread A because the GIL is disabled.
3. Thread B calls `_PyEval_EnableGIL()` and enables the GIL.
4. Thread A may briefly wait in [`tstate_try_attach()`](https://github.com/python/cpython/blob/42205143f8b3211d1392f1d9f2cf6717bdaa5b47/Python/pystate.c#L1973) depending on the timing of B's stop-the-world, but the damage is done: thread A made it past `_PyEval_AcquireLock()` and the GIL is enabled.
5. Thread A attaches, *without* holding the GIL, and continues on. Disaster ensues.

So this also needs some work.